signing/vrf.go

```diff
@@ -0,0 +1,158 @@
+package signing
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	"github.com/oasisprotocol/curve25519-voi/primitives/ed25519"
+	"github.com/oasisprotocol/curve25519-voi/primitives/ed25519/extra/ecvrf"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/datastore"
+	"github.com/spacemeshos/go-spacemesh/sql/atxs"
+)
+
+// nonceFetcher is an abstraction for VRFSigner and VRFVerifier to fetch a nonce for a given node.
+type nonceFetcher interface {
+	NonceForNode(types.NodeID) (types.VRFPostIndex, error)
+}
+
+// VRFSigner is a signer for VRF purposes.
+type VRFSigner struct {
+	privateKey []byte
+	fetcher    nonceFetcher
+
+	nodeID types.NodeID
+}
+
+// Sign signs a message for VRF purposes.
+func (s VRFSigner) Sign(msg []byte) ([]byte, error) {
+	nonce, err := s.fetcher.NonceForNode(s.nodeID)
+	if err != nil {
+		return nil, err
+	}
+
+	buf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(buf, uint64(nonce))
```

Comment:
i always get nervous when the code does this instead of using the encoding library. why not put in a private struct and use scale?
@dshulyak

Reply:
I would like to avoid `go-scale` here. `msg` is already a serialized object I'm only prefixing it with 8 bytes before I sign it. The resulting `buf` would be the same doing it like this or with `go-scale` anyway, but with go-scale I have to create another type and generate code for it to be encoded. That's a lot of extra code to just prefix a few bytes...

Reply:
the amount of code doesn't bother me. consistency across different node implementation is my main concern.
node a rust implementation cannot just rely on scale-encoding. the rust dev now has to care about what order you prefix the nonce?

having opaque []byte in the struct shouldn't be an issue for encoding.

Reply:
for now I will merge the code as is. I created an issue to come back to the encoding here: https://github.com/spacemeshos/go-spacemesh/issues/3902